### ISSUE TITLE
RFC: explicitly store cache item expiration time in persistent / database caches

### ISSUE THREAD

**Issue by holtkamp** [symfony/symfony#47868]
### Description

_TLDR: can we add a `item_expiration` column to the database table used by the [DoctrineDbalAdapter and PdoAdapter ](https://symfony.com/doc/current/components/cache/adapters/pdo_doctrine_dbal_adapter.html) to simplify maintenance on large (MySQL) database tables._

**Context**
When using the [DoctrineDbalAdapter and PdoAdapter ](https://symfony.com/doc/current/components/cache/adapters/pdo_doctrine_dbal_adapter.html) the resulting database table used for such persistent caches can become quite big: hundreds of thousands or even millions of records, depending on the application of course.

In case the lifetime of a cache item is set, it can become expired when the lifetime has "ended".

Over-time, a significant part of the persisted cache items can become "expired", only taking up space. 

The expiration time of a cache entry can be calculated using `item_time` + `item_lifetime` when `item_lifetime` is not null. 

**Challenge: remove expired entries from database table**
To clean up such tables, a naive approach would be to use a query like: 

```SQL
DELETE FROM cacheTable
  WHERE item_lifetime IS NOT NULL
  AND item_time + item_lifetime < UNIX_TIMESTAMP();
```

However, when a lot of rows are involved, this might affect the database performance, at least it does when using a MySQL database 8.0.x.

A more conservative approach would be to limit the amount of deleted rows and repeat this as long as expired items exist:

```SQL
DELETE FROM cacheTable
  WHERE item_lifetime IS NOT NULL 
  AND item_time + item_lifetime < UNIX_TIMESTAMP()
  LIMIT 1000;
SELECT SLEEP(1);
```

But I found out that with big tables this still might involve a full scan of the table which involves a lot of I/O operations, drastically slowing down the query and locking the table which should be prevented for a cache table. This seems to be expected behaviour when using the InnoDB storage engine on MySQL 8.0.x: https://bugs.mysql.com/bug.php?id=97709

**Using indices: failed**
First I tried to avoid the aforementioned problem by:
1. adding an index on the `item_time` and `item_lifetime` columns
2. adding a multi-column index named "itemTimeAndLifetime" on `(item_time, item_lifetime)`

The latter seems to improve SELECT statements which involves forced use of an index like: 

```SQL
SELECT COUNT(*) 
  FROM cacheTable FORCE INDEX (itemTimeAndLifetime) 
  WHERE item_lifetime IS NOT NULL
  AND item_time + item_lifetime < UNIX_TIMESTAMP();
```

From the documentation at https://dev.mysql.com/doc/refman/8.0/en/index-hints.html:

> Index hints work with [DELETE], but only if you use multi-table DELETE syntax like this:

```SQL
DELETE t1.* 
  FROM cacheTable t1 FORCE INDEX (itemTimeAndLifetime) 
  WHERE item_lifetime IS NOT NULL 
  AND item_time + item_lifetime < UNIX_TIMESTAMP();
```

But this also means we can not use the LIMIT clause anymore, this is an invalid query:
```SQL
DELETE t1.* 
  FROM cacheTable t1 FORCE INDEX (itemTimeAndLifetime) 
  WHERE item_lifetime IS NOT NULL 
  AND item_time + item_lifetime < UNIX_TIMESTAMP()
  LIMIT 1000;
```
When trying to combine a SELECT + DELETE, I found out that `FIND_IN_SET()` ignores indices 😞 

```SQL
SELECT GROUP_CONCAT(item_id) INTO @itemIdsToDelete
  FROM cacheTable FORCE INDEX (itemTimeAndLifetime) 
  WHERE item_lifetime IS NOT NULL
  AND item_time + item_lifetime < UNIX_TIMESTAMP() LIMIT 100;
  
DELETE FROM cacheTable WHERE FIND_IN_SET(item_id, @itemIdsToDelete);
```

**Workaround: worked**
Then I tried something else by adding a new column: `item_expiration_time` with an index on it and populated it using: 

```SQL
ALTER TABLE cacheTable
  ADD item_expiration_time INT UNSIGNED NULL,
  ADD INDEX(item_expiration_time);

UPDATE cacheTable
  SET item_expiration_time = item_time + item_lifetime
  WHERE item_expiration_time IS NULL AND item_lifetime IS NOT NULL;
```

Now cleanup operations like this are performed simple and fast:

```SQL
DELETE FROM cacheTable
  WHERE item_expiration_time < UNIX_TIMESTAMP()
  LIMIT 1000;
```

The problem is: how can we keep the `item_expiration_time` populated since the DoctrineDbalAdapter or PdoAdapter do not use this column. 

This can be done using triggers which populate the column on INSERTS or UPDATES: 

```SQL
CREATE TRIGGER setCacheExpirationTime
    BEFORE INSERT ON cacheTable
    FOR EACH ROW SET NEW.item_expiration_time = CASE WHEN NEW.item_lifetime IS NOT NULL THEN NEW.item_time + NEW.item_lifetime END;

CREATE TRIGGER updateCacheExpirationTime
    BEFORE UPDATE ON cacheTable
    FOR EACH ROW SET NEW.item_expiration_time = CASE WHEN NEW.item_lifetime IS NOT NULL THEN NEW.item_time + NEW.item_lifetime END;
```

**Request**
To simplify maintenance: would it be an idea to add the suggested `item_expiration_time` column to the DoctrineDbalAdapter and PdoAdapter to simplify maintenance on big database tables?

Of course, this _would_ imply a major change since the database schema changes...

**Comment by holtkamp** [comment]
**UPDATE**
It seems my lack of database knowledge pointed me in the wrong direction. 

Since MySQL 8.0.13 indexed expressions are supported: https://dev.mysql.com/doc/refman/8.0/en/create-index.html#create-index-functional-key-parts, also see: https://www.percona.com/blog/mysql-8-0-functional-indexes/

This allows us to create a "functional index" on the functional part `item_time + item_lifetime`: 

```sql
ALTER TABLE cacheTable ADD INDEX( (item_time + item_lifetime) );
```

With this index a DELETE statement like

```SQL
DELETE FROM cacheTable
  WHERE item_lifetime IS NOT NULL
  AND item_time + item_lifetime < UNIX_TIMESTAMP()
  LIMIT 1000;
```

has the same performance as while using the suggested new column: 

```SQL
DELETE FROM cacheTable
  WHERE item_expiration_time < UNIX_TIMESTAMP()
  LIMIT 1000;
```
Conclusion: with this approach now explicit storage expiration time in a new column `item_expiration_time` is required. 

This information might be useful to include in the documentation though?

=> closing issue.
